### PR TITLE
fix(styles): expand css shorthands in order for them to be applied properly

### LINKED ISSUE #237: CSS shorthand properties do not work well when styles are overwritten
### Steps
I'm using the `Divider` component and want to change the default CSS `margin` to another value.

In `marginStyles.tsx`, the default margin is set using `marginTop` and `marginBottom`, like this:
```jsx
const dividerStyles = {
  root: (): => ({
    marginTop: pxToRem(5 + size * 7.5),
    marginBottom: pxToRem(5 + size * 7.5),
  }),
}
```

What I did was to overwrite the margin using `margin` CSS shorthand like this:
```jsx
<Divider styles={{ root: { margin: 0 }}} />
```

### Expected Result
The `Divider` margin is overwritten to `0`.

### Actual Result
The `Divider` margin is not overwritten. We are using shorthand `margin` in order to attempt that but our mechanism merges the `Provider` `componentStyles` and the component's `styles` JS objects together so `margin` property cannot override a different property (`marginTop` or `marginBottom` in our case).

To unblock the user, the correct way to override CSS `margin` in this case is:
```jsx
<Divider styles={{ root: { marginTop: 0, marginBottom: 0 }}} />
```

However, this is undesirable and we should come up with better merging strategy for CSS in JS.

### Version
0.5.2

### PR DESCRIPTION
**Pull request by mnajdova**
Picks changes from https://github.com/stardust-ui/react/pull/296 Fixes https://github.com/stardust-ui/react/issues/237 Please see additional conversation on this PR https://github.com/stardust-ui/react/pull/296 

## Fix
This PR aims to fix #237 by expanding the css shorthands of the theme's components styles and the styles prop provided to the component while they are merged.

I added the css-shorthand-expand library and provided service for expanding the styles. There are still some points that should be addressed:

### 1. The library provides extending for the following shorthands:
```
  font
  padding
  margin
  border
  borderWidth
  borderStyle
  borderColor
  borderTop
  borderRight
  borderBottom
  borderLeft
  borderRadius
  background
  outline
```
This means that we still don't have any mechanism for expanding the other shorthands like:

```
  animation
  column-rule
  columns flex
  flex-flow grid
  grid-area
  grid-column
  grid-row
  grid-template
  list-style
  offset
  overflow
  place-content
  place-items
  place-self
  text-decoration
  transition
```
### 2. It would have been nice if we would be able to measure the performance of the components and conclude whether this changes would impact them.
I need input whether this is worth exploring further, or whether maybe we should just try to use the library when defining the styles for the themes or applying the styles prop on the components.

### Examples which are fixed with this
1. Let's say that we have a component where inside the styles prop, we are providing some extended vs shorthand styles
```
<Segment content="First Segment" styles={{marginBottom: '10px', margin: '0px'}} />
```
The example is simple, but the `margin: '0px'` can be provided based on some condition, so in some cases it would make sense.
The expected result is that the `margin: '0px'` will override the marginBottom. This PR insures that.

2. We have components that in their styles have usages of `marginRight`, `marginLeft` etc. If the user specify in the `styles` of this component (or in some nested `Provider`), that they want `margin: 0px` this should override the above specified `marginRight | Left`.
In the next example we have the `Icon` component, that by default has right margin of `8px`. Using this should override that margin.
```
<Icon name="call-video" styles={{margin: '0px'}} />
```
The prove that this didn't work before is shown in the screener changes for the icon used inside the radio item component. Basically the `radioItemGroupStyles` is overriding the right margin of the icon with the styles for the icon's slot in the following manner: `margin: '0 10px 0 0'`, but this style was never applied, and we were not even aware that the right margin was not overriden.

### CODE REVIEW

File: packages/react/src/lib/felaExpandCssShorthandsPlugin.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import * as _ from 'lodash'
+import * as _expand from 'css-shorthand-expand'
+
+// `css-shorthand-expand` is a CJS library, there are known issues with them:
+// https://github.com/rollup/rollup/issues/1267#issuecomment-446681320
+const expand = (_expand as any).default || _expand
+
+export default () => {
```
Comment (bmdalex) [inline packages/react/src/lib/felaExpandCssShorthandsPlugin.ts:33]: NIT: what about defining the method outside instead of using anonymous method?

```typescript
const expandCssShorthands = styles => {
  // function body
}

export default expandCssShorthands
```

reasons:
- more readable
- can leverage automatic import by name

File: packages/react/src/lib/felaExpandCssShorthandsPlugin.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      }
+
+      return { ...acc, [cssPropertyName]: cssPropertyValue }
+    }, {})
+  }
+
+  return expandCssShorthands
+}
+
+const transformKebabCaseKeysToCamelCase = obj => {
```
Comment (bmdalex) [inline packages/react/src/lib/felaExpandCssShorthandsPlugin.ts]: NIT: the method does not require keys to be kebab, I'd suggest:

~~`transformKebabCaseKeysToCamelCase`~~ -> `transformKeysToCamelCase` or maybe `convertKeysToCamelCase`

File: packages/react/test/specs/lib/felaExpandCssShorthandsPlugin-test.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import felaExpandCssShorthandsPlugin from 'src/lib/felaExpandCssShorthandsPlugin'
+
+const expandCssShorthands = felaExpandCssShorthandsPlugin()
+
+describe('felaExpandCssShorthandsPlugin', () => {
```
Comment (bmdalex) [inline packages/react/test/specs/lib/felaExpandCssShorthandsPlugin-test.ts:5]: nice tests 👍 couple of things:

1. I'd use `toMatchObject` instead of `toEqual` to compare objects so that we don't have to rely on key order as that is a detail of the CSS library and might change; we're also not interested in the order

2. I'd also add a test for what happens when we use an expanded prop together with its shorthand; something like:

```typescript
test('should merge expanded prop with its shorthand', () => {
  const style = {
    marginTop: '3px',
    margin: '10px',
    marginRight: '15px',
  }

  expect(expandCssShorthands(style)).toMatchObject({
    marginTop: '10px', // overridden by margin: '10px'
    marginRight: '15px', // overridden by marginRight: '15px'
    marginBottom: '10px',
    marginLeft: '10px',
  })
})
```

File: packages/react/src/lib/felaExpandCssShorthandsPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import * as _ from 'lodash'
+import * as _expand from 'css-shorthand-expand'
+
+// `css-shorthand-expand` is a CJS library, there are known issues with them:
+// https://github.com/rollup/rollup/issues/1267#issuecomment-446681320
+const expand = (_expand as any).default || _expand
+
+export default () => {
+  const expandCssShorthands = styles => {
```
Comment (bmdalex) [inline packages/react/src/lib/felaExpandCssShorthandsPlugin.ts]: NIT: type for `styles`

**Review by bmdalex — APPROVED** [review]
yuppieee 👍 few comments but looks good

File: packages/react/src/lib/felaExpandCssShorthandsPlugin.ts (hunk 1 above)
Comment (mnajdova) [reply packages/react/src/lib/felaExpandCssShorthandsPlugin.ts:33]: This is the definition that the fela renderer expect from a plugin. Other internal plugins that we have are defined in the same manner. Would rather not change at this moment.

File: packages/react/src/lib/felaExpandCssShorthandsPlugin.ts (hunk 4 above)
Comment (mnajdova) [reply packages/react/src/lib/felaExpandCssShorthandsPlugin.ts]: I checked fela's plugins, and they use here Object as a type.. Not best, but will add it..

File: packages/react/test/specs/lib/felaExpandCssShorthandsPlugin-test.ts (hunk 3 above)
Comment (mnajdova) [reply packages/react/test/specs/lib/felaExpandCssShorthandsPlugin-test.ts:5]: Added, actually this is really good case for corner cases, and testing whether one style object within itself will correctly apply the styles. Thanks for the suggestion on this one @Bugaa92

File: packages/react/src/lib/felaExpandCssShorthandsPlugin.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+export default () => {
+  const expandCssShorthands = (styles: Object) => {
+    return _.keys(styles).reduce((acc, cssPropertyName) => {
+      const cssPropertyValue = styles[cssPropertyName]
+
+      if (typeof cssPropertyValue === 'object') {
+        return { ...acc, [cssPropertyName]: expandCssShorthands(cssPropertyValue) }
+      }
+
+      const expandedProps = expand(_.kebabCase(cssPropertyName), String(cssPropertyValue))
```
Comment (kuzhelov) [inline packages/react/src/lib/felaExpandCssShorthandsPlugin.ts]: may suggest to use string interpolation approach instead of `String(cssPropertyValue)`
```js
`${cssPropertyValue}`
```

File: packages/react/src/lib/felaExpandCssShorthandsPlugin.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      if (expandedProps) {
+        return { ...acc, ...convertKeysToCamelCase(expandedProps) }
+      }
+
+      return { ...acc, [cssPropertyName]: cssPropertyValue }
```
Comment (kuzhelov) [inline packages/react/src/lib/felaExpandCssShorthandsPlugin.ts:49]: here we may simplify and use ternary expression:

```js
return { 
  ...acc,  
  ...(expandedProps 
    ? convertKeysToCamelCase(expandedProps) 
    : { [cssPropertyName]: cssPropertyValue })
}
```

File: packages/react/src/lib/felaExpandCssShorthandsPlugin.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    }, {})
+  }
+
+  return expandCssShorthands
+}
+
+const convertKeysToCamelCase = obj => {
```
Comment (kuzhelov) [inline packages/react/src/lib/felaExpandCssShorthandsPlugin.ts]: nit: this could be a one-liner

**Review by layershifter — COMMENTED** [review]
As this PR introduces a new plugin for Fela I suggest to measure performance before merge 📝

File: packages/react/src/lib/felaExpandCssShorthandsPlugin.ts (hunk 6 above)
Comment (mnajdova) [reply packages/react/src/lib/felaExpandCssShorthandsPlugin.ts:49]: There are multiple conditions there now, so I wouldn't complicate the return statements

**Review by layershifter — APPROVED** [review]
We spent some time to make it more performant, so now it's quite fast 👍
